Chatkit/Subscriptions/Subscription.swift

```diff
@@ -0,0 +1,301 @@
+import Foundation
+import class PusherPlatform.PPRequestOptions
+import enum PusherPlatform.HTTPMethod
+
+typealias SubscribeHandler = (VoidResult) -> Void
+
+protocol SubscriptionDelegate: AnyObject {
+    func subscription(_ subscription: Subscription, didReceiveEventWithJsonData jsonData: Data)
+    func subscription(_ subscription: Subscription, didReceiveError error: Error)
+}
+
+protocol Subscription {
+    func subscribe(completion: @escaping SubscribeHandler)
+    func unsubscribe()
+}
+
+class ConcreteSubscription: Subscription {
+    
+    enum State {
+        case notSubscribed
+        case subscribingStageOne(instanceWrapper: InstanceWrapper, completions: [SubscribeHandler])
+        case subscribingStageTwo(instanceWrapper: InstanceWrapper, resumableSubscription: ResumableSubscription, completions: [SubscribeHandler])
+        case subscribed(instanceWrapper: InstanceWrapper, resumableSubscription: ResumableSubscription)
+    }
+    
+    typealias Dependencies = HasInstanceWrapperFactory
+    
+    let subscriptionType: SubscriptionType // Internal `get` aids testing
```

Comment:
We should not be maintaining interface only for the sake of testing.

Reply:
See comment just below.